tsl/src/compression/compression_storage.c

```diff
@@ -0,0 +1,389 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+/*
+ * This file contains functions for manipulating compression related
+ * internal storage objects like creating the underlying tables and
+ * setting storage options.
+ */
+
+#include <postgres.h>
+#include <access/reloptions.h>
+#include <access/xact.h>
+#include <catalog/indexing.h>
+#include <catalog/objectaccess.h>
+#include <catalog/objectaddress.h>
+#include <catalog/pg_class.h>
+#include <catalog/pg_constraint.h>
+#include <catalog/toasting.h>
+#include <commands/tablecmds.h>
+#include <commands/tablespace.h>
+#include <nodes/makefuncs.h>
+#include <utils/builtins.h>
+#include <utils/lsyscache.h>
+#include <utils/syscache.h>
+
+#include "extension_constants.h"
+#include "custom_type_cache.h"
+#include "ts_catalog/catalog.h"
+#include "ts_catalog/compression_settings.h"
+#include "compression_storage.h"
+#include "hypertable.h"
+#include "compression.h"
+#include "create.h"
+#include "utils.h"
+
+#define PRINT_COMPRESSION_TABLE_NAME(buf, prefix, hypertable_id)                                   \
+	do                                                                                             \
+	{                                                                                              \
+		int ret = snprintf(buf, NAMEDATALEN, prefix, hypertable_id);                               \
+		if (ret < 0 || ret > NAMEDATALEN)                                                          \
+		{                                                                                          \
+			ereport(ERROR,                                                                         \
+					(errcode(ERRCODE_INTERNAL_ERROR),                                              \
+					 errmsg("bad compression hypertable internal name")));                         \
+		}                                                                                          \
+	} while (0);
+
+static void set_toast_tuple_target_on_chunk(Oid compressed_table_id);
+static void set_statistics_on_compressed_chunk(Oid compressed_table_id);
+static void create_compressed_chunk_indexes(Chunk *chunk, CompressionSettings *settings);
+static void clone_constraints_to_chunk(Oid ht_reloid, const Chunk *compressed_chunk);
+static List *get_fk_constraints(Oid reloid);
+
+int32
+compression_hypertable_create(Hypertable *ht, Oid owner, Oid tablespace_oid)
```

Comment:
Is this mostly code that has been moved?

Are there any changes here to be on the lookout for?

Perhaps it would be better to do refactoring separately...

Reply:
Most of this has been moved, most if the code that used to be done on the hypertable is now done per chunk and we no longer rely on inheritance propagation of columns and their attributes but now manually iterate over compressed chunks for these operations.